clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ * decode_cache.h: Library that supports caching of instruction decode
+ * information.
+ */
+
+#ifndef _DECODE_CACHE_H_
+#define _DECODE_CACHE_H_ 1
+
+#include "memref.h"
+#include "tracer/raw2trace_shared.h"
+
+#include <cstring>
+#include <mutex>
+#include <unordered_map>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Base class for storing instruction decode info. Users should sub-class this
+ * base class and implement set_decode_info_derived() to derive and store the decode
+ * info they need.
+ */
+class decode_info_base_t {
+public:
+    /**
+     * Sets the decode info for the provided \p instr which was allocated using the
+     * provided \p dcontext for the provided \p memref_instr. This is done using
+     * the set_decode_info_derived() provided by the derived class. Additionally,
+     * this does other required bookkeeping.
```

Comment:
Based on the ownership comment below, this takes over ownership of "instr"?  Add to comment.